### PR TITLE
James 1858: Docker image on Guice+JPA with initial data

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: dockerfiles/compilation/java-8/compile.sh [hunk 1]
```diff
@@ -44,17 +45,27 @@ git checkout $SHA1
...
 
 # Retrieve result
 
 if [ $? -eq 0 ]; then
-   cp server/app/target/james-server-app-*-app.zip $DESTINATION
-   cp server/container/guice/cassandra-guice/target/james-server-cassandra-guice-*-SNAPSHOT.jar $DESTINATION
-   cp -r server/container/guice/cassandra-guice/target/james-server-cassandra-guice-*-SNAPSHOT.lib $DESTINATION
-   cp server/container/cli/target/james-server-cli-*.jar $DESTINATION
-   cp -r server/container/cli/target/james-server-cli-*.lib $DESTINATION
+   if [ -d "$CASSANDRA_DESTINATION" ]; then
```
Comment (mbaechler) [inline dockerfiles/compilation/java-8/compile.sh]: while we are at it, we could remove the version from artifact names so we don't have to change all dockerfile when we change version string.

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java [hunk 2]
```diff
@@ -102,6 +102,7 @@ public static void main(String[] args) {
         } catch (IOException ioe) {
             failWithMessage("Error connecting to remote JMX agent : " + ioe.getMessage());
         } catch (Exception e) {
+            e.printStackTrace();
```
Comment (mbaechler) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: don't we have a logging framework ?

File: dockerfiles/run/guice/jpa/destination/conf/log4j.properties [hunk 3]
```diff
@@ -0,0 +1,137 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
```
Comment (chibenwa) [inline dockerfiles/run/guice/jpa/destination/conf/log4j.properties]: Is this file needed ? Don't we use logback ?

File: dockerfiles/run/guice/jpa/destination/conf/sqlResources.xml [hunk 4]
```diff
@@ -0,0 +1,931 @@
+<!--
```
Comment (chibenwa) [inline dockerfiles/run/guice/jpa/destination/conf/sqlResources.xml]: Is it needed ?

File: dockerfiles/run/guice/provisioned/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Run James
```
Comment (chibenwa) [inline dockerfiles/run/guice/provisioned/Dockerfile]: I think we should have only one provisionned images, built on top of guice

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java (hunk 2 above)
Comment (quynhn) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: OK, I changed to use the logger instead of

File: dockerfiles/run/guice/provisioned/Dockerfile (hunk 5 above)
Comment (quynhn) [inline dockerfiles/run/guice/provisioned/Dockerfile]: Done

File: dockerfiles/run/guice/jpa/destination/conf/sqlResources.xml (hunk 4 above)
Comment (quynhn) [inline dockerfiles/run/guice/jpa/destination/conf/sqlResources.xml]: Remove

File: dockerfiles/run/guice/jpa/destination/conf/log4j.properties (hunk 3 above)
Comment (quynhn) [inline dockerfiles/run/guice/jpa/destination/conf/log4j.properties]: Done

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java [hunk 6]
```diff
@@ -102,7 +105,7 @@ public static void main(String[] args) {
         } catch (IOException ioe) {
             failWithMessage("Error connecting to remote JMX agent : " + ioe.getMessage());
         } catch (Exception e) {
-            e.printStackTrace();
+            LOG.error("Error on command: ", e);
             failWithMessage("Error while executing command:" + e.getMessage());
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: Maybe we can get a better user experiece with : 

```
failWithMessage("Error " + e.getClass() + " while executing command:" + e.getMessage());
```

**Comment by chibenwa** [comment]
test this please

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java (hunk 6 above)
Comment (quynhn) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: As I traced on that, the message is not null. The exception class is NullPointerException but on another part. So think the exception class is not enough. I changed a bit to print out stasktrace in the log file.

Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: I'm not asking for removing the log, I'm asking for making the message returned to the user a bit more specific.

About empty messages, maybe you did not encountered it, but I did ;-).

Comment (quynhn) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: OK, got it and I changed

**Comment by chibenwa** [comment]
I'm ok when you will demonstrate it works on CI ^^

**Comment by chibenwa** [comment]
https://james.open-paas.org/job/QuynhTestOnMyBranch/51/ is green, you have my :+1:

File: server/container/guice/jpa-guice/pom.xml [hunk 7]
```diff
@@ -194,7 +194,7 @@
                  <dependency>
                     <groupId>org.apache.derby</groupId>
                     <artifactId>derby</artifactId>
-                    <scope>test</scope>
+                     <scope>compile</scope>
```
Comment (aduprat) [inline server/container/guice/jpa-guice/pom.xml]: I'm not sure about this modification, you are embedding Derby for every builds of jpa-guice; but the user may use another database.
Derby should come in another docker container (such as Cassandra).

Comment (quynhn) [inline server/container/guice/jpa-guice/pom.xml]: Yes, if user would like to use another database, so they need to change property file (james-database.properties) with the right driver, url ... and they need jar package for that driver.
In our case, we have to launch server and provision some initial data.
And we have driver class is: database.driverClassName=org.apache.derby.jdbc.EmbeddedDriver
So I think it need to modify that scope so that the jar file is included when we package server.
Is that correct?

Comment (chibenwa) [inline server/container/guice/jpa-guice/pom.xml]: I do not think this is correct. They would need to recompile JPA Guice with there driver added (so that it is part of the classpath).

Then they would need to add specific driver configuration, I guess in james-database.properties.
james-database

Comment (quynhn) [inline server/container/guice/jpa-guice/pom.xml]: OK, let me try on my local

**Comment by quynhn** [comment]
test this please
